File: content/problems/0057-insert-interval/solutions/three-phase-scan/Solution.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> answer = new ArrayList<>(); int index = 0;
        while (index < intervals.length && intervals[index][1] < newInterval[0]) answer.add(intervals[index++]);
        while (index < intervals.length && intervals[index][0] <= newInterval[1]) { newInterval[0] = Math.min(newInterval[0], intervals[index][0]); newInterval[1] = Math.max(newInterval[1], intervals[index++][1]); }
        answer.add(newInterval); while (index < intervals.length) answer.add(intervals[index++]);
        return answer.toArray(new int[answer.size()][]);
    }
}

File: content/problems/0057-insert-interval/solutions/three-phase-scan/solution.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> answer; int index = 0;
        while (index < (int)intervals.size() && intervals[index][1] < newInterval[0]) answer.push_back(intervals[index++]);
        while (index < (int)intervals.size() && intervals[index][0] <= newInterval[1]) { newInterval[0] = min(newInterval[0], intervals[index][0]); newInterval[1] = max(newInterval[1], intervals[index++][1]); }
        answer.push_back(newInterval); while (index < (int)intervals.size()) answer.push_back(intervals[index++]); return answer;
    }
};
